shared/src/main/kotlin/org/javacs/kt/util/OSContext.kt

```diff
@@ -0,0 +1,27 @@
+package org.javacs.kt.util
+
+/**
+ * Tasks that depends on user's OS
+ */
+interface OSContext {
+    /**
+     * Suggests the candidate locations of the given JAR
+     *
+     * @param name the name of the JAR
+     * @return the candidate full paths to the JAR
+     */
+    fun candidateAlternativeLibraryLocations(name: String): Array<String>
```

Comment:
I am not sure if I really understand this abstraction.

While it's nice to have the implementation separated by OS, this is fundamentally pretty specific to a certain kind of (fallback) library resolution, that shouldn't (IMO) live under a name as generic as `org.javacs.kt.util.OSContext`. I could imagine fitting `ShellPathUtils` into an interface like this, but that's a different use-case.

I think, in this case specifically I would probably even drop the interface/classes entirely and just inline the `when`-check over the OS with the corresponding paths into `BackupClassPathResolver`'s `findAlternativeLibraryLocation`.

Alternatively, if the `BackupClassPathResolver` is getting too large, we might want to add a `SnapClassPathResolver`/`ScoopClassPathResolver`?

Reply:
>  this is fundamentally pretty specific to a certain kind of (fallback) library resolution, that shouldn't (IMO) live under a name as generic as org.javacs.kt.util.OSContext. I could imagine fitting ShellPathUtils into an interface like this, but that's a different use-case.

I thought other methods that depends on the type of OS may be added in the future. I'll have to change their names.

> I think, in this case specifically I would probably even drop the interface/classes entirely and just inline the when-check over the OS with the corresponding paths into BackupClassPathResolver's findAlternativeLibraryLocation.

I oppose it if it's called more than once through the entire of the program. We have to choose the appropriate instance according to user's OS. That that wasn't chosen mustn't be loaded.
The `USERPROFILE` value is used only in Windows, so it must be a state enclosed in the Windows class.
If you won't use interfaces, 

> we might want to add a SnapClassPathResolver/ScoopClassPathResolver?

Do they mean singletons (`object`s) that implement `ClassPathResolver`?

Reply:
@tats-u , I see that you have not been getting a reply here 🙁 I'm pretty sure that @fwcd meant creating singletons like you are asking.